### PR TITLE
feat: automatically expose connection via Rise Proxy

### PR DESCRIPTION
**Pull request by grabbou**
Given time constraints, this is much simpler and smaller implementation of original Tunnel idea we had in mind.

Key highlights:
- DigitalOcean VPS running [Sish](https://github.com/antoniomika/sish) via Docker Compose (I took me a while to figure this out). We're on $6 plan a month. We will see how long it lasts.
- We use [SSH tunneling](https://www.ssh.com/academy/ssh/tunneling)
- We use `*.proxy.rise.tools` subdomain

Benefits:
- No additional tools required. Just SSH client.
- No cloud service involved. Less things to maintain. We will eventually need it (more on it later).

How this works:
- Project is assigned its unique ID (via `uuid.v4()`) 
- Project ID is stored in `.rise/projectId` file
- Tunnel will be open on each server start automatically, with 5s timeout 
- Hostname is fixed to be <projectId>.proxy.rise.tools.

Known limitations:
- URL must be unique, so we use rather ugly projectId.proxy.rise.tools. It's not that big of a deal, since you scan QR Code or copy the deep link
- Anyone can start the tunnel knowing your projectId. It means that it's not 100% secure (although brute force guessing your projectId is only theoretically possible). However, if you share your connection with network, it may become more likely. **That's why tunnelling should only be used for development**

[All these two limitations can be fixed easily](https://github.com/antoniomika/sish/issues/313) - I opened up an issue in the project we're using.

Once this is implemented, we will be able to:
- use nicer names for projects, e.g. "fancy-project"
- allow users to set custom names

**Preview:**
Screenshot 2024-07-21 at 02 03 32

TBD:
- Better UX if tunnel fails and while tunnel is loading (and it's been more than 2s already)
- Warn if no `''` model is present. QR Code points to index by default and this fails otherwise. This should be configurable.
- Provide default screen for example. When users add it, they will loose default Rise Playground screen with list of available examples. We should print it there instead.
- ONE DAY: maybe that default Playground screen can be Rise too!
- SSL

### CODE REVIEW

**Comment by arpitBhalla** [comment]
Does watching files attempt to reconnect the proxy?

**Comment by grabbou** [comment]
> Does watching files attempt to reconnect the proxy?

Restarting server will simply restart the session.

We should likely decouple it, so we don't connect every time you save things. But this may require a bit more clever design - open to suggestions how this can be solved

**Comment by arpitBhalla** [comment]
how about we create a `.pid` file to store the SSH process ID and run SSH in the background? On restart, we can check if SSH is still running with that process ID. If not, we can create a new SSH process.

**Comment by grabbou** [comment]
> how about we create a `.pid` file to store the SSH process ID and run SSH in the background `{detached: true}`? On restart, we can check if SSH is still running with that process ID. If not, we can create a new SSH process.

Happy to accept a follow-up PR that does just that. I think that would be useful. No need to pollute our server with multiple SSH requests if we're reloading often.

Question: how would we clean this tunnel automatically if our server crashed and never restarted? Any way to make it close automatically?

**Comment by arpitBhalla** [comment]
I think the options `ServerAliveInterval`, `ServerAliveCountMax`, and `ExitOnForwardFailure` might work. What do you think?

**Comment by grabbou** [comment]
I ended up coming with much better, cloud-less solution. On new connection, on the server, when you create tunnel URL, do this:
- take requested user subdomain
- take user public fingerprint
- hash them

this is your URL. Return this URL to the client and use as a tunnel

Code:
```
hash := md5.New()
hash.Write([]byte(addr + sshConn.SSHConn.Permissions.Extensions["pubKeyFingerprint"]))
hashedAddr := fmt.Sprintf("%x", hash.Sum(nil))
proposedHost = fmt.Sprintf("%s%s.%s", hashedAddr, hostExtension, viper.GetString("domain")) 
```

This PR will remain unaffected. All changes will be done on the server.

**Comment by grabbou** [comment]
Okay, I managed to modify the source of the proxy so that it hashes:

Log from the server
> Trying access: 0ba3656-dd69-40ae-b64d-49b2b8e3dd54, from fingerprint: SHA256:wCM0tvO/WJg+OkfaayQAKFAn7yT6Viw2w3LiG82QzWg, hash: f05ec229d85d7bafd545f0bbaeafffc9

The URL:

http://f05ec229d85d7bafd545f0bbaeafffc9.proxy.rise.tools/ 

URL is a hash of authorised public key fingerprint and passed projectId. 

TL;DR

only users that have your private/public key will be able to overwrite tunnel URL, which is expected.

**Comment by arpitBhalla** [comment]
I absolutely loved how it turned out. You are a genius!

**Comment by grabbou** [comment]
that's what team work is! discussions make it

**Comment by grabbou** [comment]
In a follow-up, we need to change the behavior when connection fails.

Instead of throwing an error and quitting the process (after timeout of 5s), we should start normal server (without tunnel) instead. It might be likely that our tunnel goes down and we don't want everyone to not be able to develop with Rise. 

We should also print out an error (in grey color) that there was an error connecting and remote preview is not available.

We should also think about flag to disable/enable this behavior programatically.
